packages/ember-data/lib/system/record-array-manager.js

```diff
@@ -60,7 +60,8 @@ export default Ember.Object.extend({
   */
   updateRecordArrays: function() {
     forEach(this.changedRecords, function(record) {
-      if (record.isDeleted()) {
+      // TODO: it can be refactored after #2862 && #2859 are closed
+      if (get(record, 'record.isDestroyed') || get(record, 'record.isDestroying') || (record.isDeleted() && !record.isDirty())) {
```

Comment:
@igor @drogus I'm still having trouble figuring out how to form this conditional, but only 4 failing tests now...